Cleanup persistence layer

richardbizik opened the pull request:
this PR contains part 3 work on #40
closes #40

nitram509 reviewed (approved):
Very large PR - would wish smaller ones ... e.g. the replacement for gocrest to testtify would be a prime candidate.

nitram509 commented on pkg/bpmn/engine.go:
```diff
@@ -162,73 +281,101 @@ func (state *Engine) run(instance *processInstanceInfo) (err error) {
+			err = tCmd.err
+			// TODO: do something with the error?
+			instance.State = runtime.ActivityStateFailed
+			break mainLoop
```
That seems wrong to me.
AFAIK, when an Activity goes Failed, other parallel activities should continue, and breaking the main loop prevents that.

richardbizik commented:
> Very large PR - would wish smaller ones ... e.g. the replacement for gocrest to testtify would be a prime candidate.

I started replacing gocrest only in the files I touched and had to write tests in. But as of now there are only few files remaining so I might complete that replacement.

nitram509 commented:
> > Very large PR - would wish smaller ones ... e.g. the replacement for gocrest to testtify would be a prime candidate.
> 
> I started replacing gocrest only in the files I touched and had to write tests in. But as of now there are only few files remaining so I might complete that replacement.

Understood.
Considering the "low capacity" and early stage of the project ... I would be totally fine in merging this and continue with new PR for such ... or leave it for later.

richardbizik commented:
I added the changes we talked about with @dmytro-gundartsev, added tests for the storage interface. Both in-memory and rqlite are now using the same set of tests.
And updated schema to have more information about things stored in the columns.
@nitram509 I also removed the goto from main loop

dmytro-gundartsev commented on internal/cluster/partition_persistence.go:
```diff
@@ -0,0 +1,723 @@
+			TimerState:           runtime.TimerState(timer.State),
+			CreatedAt:            time.UnixMilli(timer.CreatedAt),
+			DueAt:                time.UnixMilli(timer.DueAt),
+			Duration:             time.Millisecond * time.Duration(timer.DueAt-timer.CreatedAt),
```
Nice trick to not store Duration ;).
Please check if the duration should be in Millies or seconds (in engine.createTimer we use time.Seconds)

richardbizik replied:
we use time.Duration (which is unit agnostic)

dmytro-gundartsev replied:
Was confused. Now clear. It has to be  all the time adjust to the nanosec counts. Thank you for the explanation

dmytro-gundartsev commented on internal/cluster/partition_persistence.go:
```diff
@@ -0,0 +1,723 @@
+
+func SaveTimerWith(ctx context.Context, db *sql.Queries, timer runtime.Timer) error {
+	err := db.SaveTimer(ctx, sql.SaveTimerParams{
+		Key:                  timer.GetKey(),
```
Key and ElementInstanceKey have the same value (taken by interface method GetKey() and directly struct field Key. 
Do we need both of them?

richardbizik replied:
I would leave this for later...dont want to add another 1000 lines to this PR
I think that the element_instance_key should be more like origin_acitivity_key but the fix should not be part of this PR

dmytro-gundartsev commented on internal/cluster/partition_persistence.go:
```diff
@@ -0,0 +1,723 @@
+func (rq *RqLiteDB) SaveJob(ctx context.Context, job runtime.Job) error {
+	err := SaveJobWith(ctx, rq.queries, job)
+	if err != nil {
+		rows, err := rq.QueryContext(ctx, "select * from process_instance")
```
Do we really need to query all the process instances and have them deserialised in the objects?
And what is the purpose of that action to cause panic (like general RQLite storage liveness detection)?

dmytro-gundartsev reviewed (commented):
Well done! Just several unclear points (please see the inline comments). Have analysed only latest (today's) commits
